### PR TITLE
Bibtex integration for document loader and retriever

### PR DESCRIPTION
**Pull request by wavefrontshaping**
# Bibtex integration

Wrap bibtexparser to retrieve a list of docs from a bibtex file.
* Get the metadata from the bibtex entries
* `page_content` get from the local pdf referenced in the `file` field of the bibtex entry using `pymupdf`
* If no valid pdf file, `page_content` set to the `abstract` field of the bibtex entry
* Support Zotero flavour using regex to get the file path
* Added usage example in `docs/modules/indexes/document_loaders/examples/bibtex.ipynb`

## Who can review?

My best guess: @eyurtsev, @dev2049

### CODE REVIEW

**Comment by wavefrontshaping** [comment]
> @wavefrontshaping do you have any unit tests you could include?

@eyurtsev yep, I added a simple unit test for the loader.

File: langchain/retrievers/bibtex.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+
+class BibtexRetriever(BaseRetriever, BibtexparserWrapper):
+    """
+    It is effectively a wrapper for BibtexparserWrapper.
+    It wraps load() to get_relevant_documents().
+    It uses all BibtexparserWrapper arguments without any change.
+    """
+
+    def get_relevant_documents(self, file_path: str) -> List[Document]:
```
Comment (dev2049) [inline langchain/retrievers/bibtex.py:14]: get_relevant_documents should only take in a query (shouldn't change the BaseRetriever interface). see retrievers/arxiv.py for an example

Comment (dev2049) [reply langchain/retrievers/bibtex.py:14]: if we don't have a good way of searching over bibtex we can remove the Retriever for now, just a document loader is already nice

Comment (wavefrontshaping) [reply langchain/retrievers/bibtex.py]: Fair point, I removed the retriever.

File: langchain/utilities/bibtex.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    def _meta_str(self, entry: Any) -> str:
+        """
+        Return the meta information of an article in string format.
+        """
+        metadata = self._get_meta(entry)
+        return "\n".join([f"{key}: {value}" for key, value in metadata.items()])
+
+    def _get_metata(self, entry: Any) -> str:
+        meta = {
+            "ID": entry.get("ID", ""),
```
Comment (eyurtsev) [inline langchain/utilities/bibtex.py:74]: @dev2049  we'll use snake-case here

File: langchain/utilities/bibtex.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            "series",
+            "type",
+            "doi",
+            "issn",
+            "isbn",
+        ]
+        if self.load_all_available_meta:
+            for field in optional_fields:
+                if field in entry:
+                    meta[field.capitalize()] = entry.get(field)
```
Comment (eyurtsev) [inline langchain/utilities/bibtex.py:106]: Will break, but reminding to change to snake_case 

```suggestion
                    meta[field.snake_case()] = entry.get(field)
```

File: langchain/utilities/bibtex.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    def load(self, file_path: str) -> List[Document]:
+        """
+        Load bibtex file using bibtexparser and get the article texts plus the
+        article metadata.
+        See https://bibtexparser.readthedocs.io/en/master/
+
+        Returns: a list of documents with the document.page_content in text format
+
+        """
+        print("Running bibtexparser wrapper LOADER")
```
Comment (eyurtsev) [inline langchain/utilities/bibtex.py:139]:
```suggestion
```

File: langchain/utilities/bibtex.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        try:
+            import fitz
```
Comment (eyurtsev) [inline langchain/utilities/bibtex.py:141]: We'll need to swap out the default parser implementation here so fitz is not hard-coded

**Comment by eyurtsev** [comment]
@wavefrontshaping one of the maintainers can merge this from this point given that there's a test, we'll likely change a bit how the metadata looks like and how the source file is parsed.

One concern about behavior is that the pdf paper is used when available and if not falling back silently to the abstract, but the user doesn't have a way of configuring the behavior (e.g., always use abstract, or skip over entries where underlying document isn't available)

**Comment by wavefrontshaping** [comment]
> One concern about behavior is that the pdf paper is used when available and if not falling back silently to the abstract, but the user doesn't have a way of configuring the behavior (e.g., always use abstract, or skip over entries where underlying document isn't available)

I see your point. More control from the user would indeed be better. 
However, I do not know the best way to address this.
My point being, let's say you have an option like `content_source` with choices:
* `abstract` not to look at files and always get the abstract, ok,
* `file_or_abstract`, the current behavior, abstract as a fallback content if not usable file,
*  `file` to retrieve the local pdf files. But what happens, if there is no file or no valid one? Do we want to raise an error, which may be detrimental as in a long list of references, it is highly likely that some would fail. Or do we want to silently go to an empty string or skip this entry, which poses the same problem as the original one.

Is there a defined strategy for similar issues in other loaders?

**Comment by wavefrontshaping** [comment]
Any thoughts?

**Comment by eyurtsev** [comment]
@wavefrontshaping thanks for the ping! 

We can keep the current auto behavior and enhance with more loading strategies after landing this PR. Not a blocker IMO.

-----

My guess is that users that choose `file` will probably be OK skipping entries were the file cannot be loading (maybe with a potential warning or error logged). However, we could add another file loading strategy that raises on error. Anyway, I think we can add those as needed via the __init__ without causing any breaking changes.

**Comment by eyurtsev** [comment]
@wavefrontshaping Going to resolve lint conflicts and merge

**Comment by eyurtsev** [comment]
@wavefrontshaping merging here: https://github.com/hwchase17/langchain/pull/5137 

Resolved linting and type issues and did some refactors

**Comment by eyurtsev** [comment]
Closing as this has been moved to a separate PR. @wavefrontshaping questions on other PR for you

**Pull request closed without merging**